src/category_theory/single_obj.lean

```diff
@@ -0,0 +1,122 @@
+`single_obj.to_End`.
+
+## Implementation notes
+
+- `category_struct.comp` on `End (single_obj.star α)` is `flip (*)`, not `(*)`. This way
+  multiplication on `End` agrees with the multiplication on `α`.
+
+- By default, Lean puts instances into `category_theory` namespace instead of
+  `category_theory.single_obj`, so we give all names explicitly.
+-/
+import category_theory.endomorphism category_theory.groupoid category_theory.Cat
+import data.equiv.algebra algebra.Mon.basic
+import tactic.find
+
+universes u v w
+
+namespace category_theory
+/-- Type tag on `unit` used to define single-object categories and groupoids. -/
+def single_obj (α : Type u) : Type := unit
+
+namespace single_obj
+
+variables (α : Type u)
+
+/-- One and `flip (*)` become `id` and `comp` for morphisms of the single object category. -/
+instance category_struct [has_one α] [has_mul α] : category_struct (single_obj α) :=
+{ hom := λ _ _, α,
+  comp := λ _ _ _ x y, y * x,
+  id := λ _, 1 }
+
+/-- Monoid laws become category laws for the single object category. -/
+instance category [monoid α] : category (single_obj α) :=
+{ comp_id' := λ _ _, one_mul,
+  id_comp' := λ _ _, mul_one,
+  assoc' := λ _ _ _ _ x y z, (mul_assoc z y x).symm }
+
+/-- Groupoid structure on `single_obj α` -/
+instance groupoid [group α] : groupoid (single_obj α) :=
+{ inv := λ _ _ x, x⁻¹,
+  inv_comp' := λ _ _, mul_right_inv,
+  comp_inv' := λ _ _, mul_left_inv }
+
+protected def star : single_obj α := unit.star
+
+/-- The endomorphisms monoid of the only object in `single_obj α` is equivalent to the original
+     monoid α. -/
+def to_End_equiv [monoid α] : End (single_obj.star α) ≃* α := mul_equiv.refl α
+
+/-- Reinterpret an element of a monoid as an element of the endomorphisms monoid of the only object
+    in the `single_obj α` category. -/
+def to_End {α} [monoid α] (x : α) : End (single_obj.star α) := x
+
+lemma to_End_def [monoid α] (x : α) : to_End x = x := rfl
+
+/-- There is a 1-1 correspondence between monoid homomorphisms `α → β` and functors between the
+    corresponding single-object categories. It means that `single_obj` is a fully faithful
+    functor. -/
+def map_hom_equiv {α : Type u} {β : Type v} [monoid α] [monoid β] :
+  { f : α → β // is_monoid_hom f } ≃ (single_obj α) ⥤ (single_obj β) :=
+{ to_fun := λ f,
+  { obj := id,
+    map := λ _ _, f.1,
+    map_id' := λ _, f.2.map_one,
+    map_comp' := λ _ _ _ x y, @is_mul_hom.map_mul _ _ _ _ _ f.2.1 y x },
+  inv_fun := λ f, ⟨@functor.map _ _ _ _ f (single_obj.star α) (single_obj.star α),
+    { map_mul := λ x y, f.map_comp y x, map_one := f.map_id _ }⟩,
+  left_inv := λ ⟨f, hf⟩, rfl,
+  right_inv := assume f, by rcases f; obviously }
+
+/-- Reinterpret a monoid homomorphism `f : α → β` as a functor `(single_obj α) ⥤ (single_obj β)`.
+See also `map_hom_equiv` for an equivalence between these types. -/
+def map_hom {α : Type u} {β : Type v} [monoid α] [monoid β] (f : α → β) [hf : is_monoid_hom f] :
```

Comment:
Perhaps add a simp lemma relating this function with the coercion of `map_hom_equiv` to a function.

Reply:
(Or remove this definition and state the lemmas directly about the coercion?)

Reply:
@semorrison If I write `map_hom_equiv f`, Lean fails to coerce `map_hom_equiv` to a function. I have no idea why.

Reply:
@digama0 Why do such coercions fail so often?

Reply:
@jcommelin I made `α` and `β` explicit for `map_hom_equiv`, and now it works.

Reply:
@semorrison Marked definition of `map_hom` as `@[reducible]`. Should I use `@[simp]` instead? Or in addition to `@[reducible]`?

Reply:
I think `@[reducible]` should be okay.